project-management/src/main/java/life/qbic/projectmanagement/application/sample/SampleRegistrationServiceV2.java

```diff
@@ -42,32 +51,46 @@ public class SampleRegistrationServiceV2 {
   private final BatchRepository batchRepository;
   private final SampleCodeService sampleCodeService;
   private final DeletionService deletionService;
+  private final ConfoundingVariableService confoundingVariableService;
 
   @Autowired
   public SampleRegistrationServiceV2(BatchRegistrationService batchRegistrationService,
       SampleRepository sampleRepository, BatchRepository batchRepository,
       SampleCodeService sampleCodeService,
-      DeletionService deletionService) {
+      DeletionService deletionService, ConfoundingVariableService confoundingVariableService) {
     this.batchRegistrationService = Objects.requireNonNull(batchRegistrationService);
     this.sampleRepository = Objects.requireNonNull(sampleRepository);
     this.batchRepository = Objects.requireNonNull(batchRepository);
     this.sampleCodeService = Objects.requireNonNull(sampleCodeService);
     this.deletionService = Objects.requireNonNull(deletionService);
+    this.confoundingVariableService = confoundingVariableService;
   }
 
   @PreAuthorize("hasPermission(#projectId, 'life.qbic.projectmanagement.domain.model.project.Project', 'WRITE')")
   @Async
+  @Transactional(propagation = Propagation.REQUIRES_NEW)
   public CompletableFuture<Void> registerSamples(Collection<SampleMetadata> sampleMetadata,
-      ProjectId projectId, String batchLabel, boolean batchIsPilot, ExperimentId experimentId)
+      ProjectId projectId, String batchLabel, boolean batchIsPilot, ExperimentReference experiment)
       throws RegistrationException {
     var result = batchRegistrationService.registerBatch(batchLabel, batchIsPilot, projectId,
-        experimentId);
+        ExperimentId.parse(experiment.id()));
     if (result.isError()) {
       throw new RegistrationException("Batch registration failed");
     }
     var batchId = result.getValue();
     try {
-      var sampleIds = registerSamples(sampleMetadata, batchId, projectId);
+      var registeredSamples = registerSamples(sampleMetadata, batchId, projectId);
+      for (Entry<Sample, SampleMetadata> registeredSample : registeredSamples.entrySet()) {
+        Map<VariableReference, String> levels = new HashMap<>();
+        registeredSample.getValue().confoundingVariables().forEach((k, v) -> levels.put(k.id(), v));
```

Comment:
suggestion: you can use `Collectors.toMap()` here, makes it more concise.

Reply:
To use `Collectors#toMap`, I need to create a stream first. So it does not make it more readable I feel. However I have no strong opinion on the matter. So if you think the lower example is better I will change it.
```
confoundingVariables().foreach((k, v) -> levels.put(k.id(), v));
```
```
registeredSample.getValue().confoundingVariables()
            .entrySet()
            .stream()
            .collect(Collectors.toMap(
                entry -> entry.getKey().id(),
                Entry::getValue));
```

Reply:
Yes, just nit, but to be fair it is:

```java
Map<VariableReference, String> levels = new HashMap<>();
registeredSample.getValue().confoundingVariables().forEach((k, v) -> levels.put(k.id(), v));
```

vs 

```java
var levels = registeredSample.getValue().confoundingVariables()
                                .entrySet().stream()
                                .collect(Collectors.toMap(entry -> entry.getKey().id(), Entry::getValue));

The last one is side-effect free, vs the first is depending on an external map initialization :P